### ISSUE TITLE
Very slow Syncnet training proccess

### ISSUE THREAD

**Issue by NikitaKononov** [primepake/wav2lip_288x288#19]
Hello!
Did you modified data loader? Because this implementation have some speed troubles
48h training and loss dropped from 0.7 to 0.6 
There've been 94 epochs (~ 29.5k steps). It's kinda slow... I think the problem is in dataloader. 
On GPU Tesla V100

My params:
lr 1e-5
batch size 128
loss func nn.BCEWithLogitsLoss()

Dataset has 40k vids
All videos are filtered, sync corrected, have only one face with nice resolution 

Can you please give any suggestions?

**Comment by NikitaKononov** [comment]
Maybe torch dataloader is not optimal?

**Comment by primepake** [comment]
check this thread https://github.com/Rudrabha/Wav2Lip/pull/265

**Comment by NikitaKononov** [comment]
> check this threa

Seen it, but it doesn't fit for large dataset
My preprocessed data is ~190GB :(

**Comment by NikitaKononov** [comment]
> check this thread [Rudrabha/Wav2Lip#265](https://github.com/Rudrabha/Wav2Lip/pull/265)

You did same changes while your training? Or you didn't modify loader?

**Comment by primepake** [comment]
you can extract audio to spectoram as NumPy file and then save it to storage. It will optimize your time training

**Comment by NikitaKononov** [comment]
> you can extract audio to spectoram as NumPy file and then save it to storage. It will optimize your time training

Thanks!
Are there any other optimising tips? Maybe to modify while 1 cycle to work better?

**Comment by lsw5835** [comment]
Hi @primepake , Did you changes lossfunction to nn.BCEWithLogitsLoss()?

**Comment by primepake** [comment]
I need to customize your loss function into range [0,1]

**Comment by hannarud** [comment]
@NikitaKononov did you manage to speedup the training? I train on AVSpeech, and only 100K videos are used currently, but my training speed is 500 steps/hour. I'm not sure until which epoch I'll need to train (because the loss is stuck at 0.69 currently), but seems like at least until 350K, which means it'll take me at least 30 days Oo

**Comment by primepake** [comment]
hey! the problem 0.69 loss due to your dataset, you need to filter it carefully

**Comment by hannarud** [comment]
Hi @primepake @NikitaKononov!
I wonder how many epochs did your lipsync expert training took in the end? I guess I managed to speed up each epoch time, and the loss even starts to drop below 0.69 (after enormous 800000 steps, but still).
So do you recall, how much steps/epochs was required? Several millions? Or my training is in general goes bad? I'm now running at 1600000 epochs and the loss is 0.39 and it still goes down but veeeeery slowly.

**Comment by NikitaKononov** [comment]
> Hi @primepake @NikitaKononov! I wonder how many epochs did your lipsync expert training took in the end? I guess I managed to speed up each epoch time, and the loss even starts to drop below 0.69 (after enormous 800000 steps, but still). So do you recall, how much steps/epochs was required? Several millions? Or my training is in general goes bad? I'm now running at 1600000 epochs and the loss is 0.39 and it still goes down but veeeeery slowly.

Hello! My current SN model is on 229007 step (516 epoch) and has train loss ~0.38
Batch size 256
Yeah, I am still encountering same problem with frozen loss on 0.35-0.4 values... Maybe scheduled LR may help us

**Comment by NikitaKononov** [comment]
While of freezing on 0.35-0.4 loss model starts to behave slightly overfitted (evaluation loss starts to jump up and down)
I've enlarged my dataset by almost 50% (93k vids before, 139k after), but it seems to have no effect. 
I'll continue my research

**Comment by hannarud** [comment]
@NikitaKononov thank you for the response!
Actually, when I saw the overfitting (training loss was around 0.35-0.34-0.33, whle validation stayed at 0.38-0.39-0.40), I went back to the step where no overfitting was observed and re-started from there with a smaller learning rate. That was on loss 0.39, step 980000 (I didn't keep track of epoch number, need to check it). Now the loss is 0.37 and no overfitting yet, but goes down very slowly (I think, it was 0.38 yesterday and 0.37 today).
But well, my step numbers are really big compared to yours...

My dataset is 100k small length videos, and yes, I thought about enlarging it, but didn't do it yet.

**Comment by hannarud** [comment]
@NikitaKononov and do you find the quality on 0.35 satisfactory for the main wav2lip training?

**Comment by NikitaKononov** [comment]
@hannarud 
I've had same situation
If you use 1-e4 or 1-e3 LR at the beginning, you'll get smth around 0.45 kina fast

step numbers depends on batchsize ) if I've used BS 16 instead of 256, I would have got 16x more steps )

**Comment by NikitaKononov** [comment]
> @NikitaKononov and do you find the quality on 0.35 satisfactory for the main wav2lip training?

Can't answer this question yet
In my latest attempt of W2L training I've used syncnet with relu convs (now I use Leaky) trained on 1.75x smaller dataset
And val loss of syncnet was 0.39
Results were not ideal, only 50-60% lip movements seemed to be accurate

I think we should try to use codes from original W2L repo (eval or test directory, don't remember exactly) as an extra way to benchmark model
